src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/ordinal/OrdinalRewriterRule.java

```diff
@@ -0,0 +1,210 @@
+    private RewriteRuleExecutor<SQLQueryExpr> ruleExecutor;
+
+    public OrdinalRewriterRule(String sql) {
+        this.sql = sql;
+    }
+
+    @Override
+    public boolean match(SQLQueryExpr root) throws SQLFeatureNotSupportedException {
+        SQLSelectQuery sqlSelectQuery = root.getSubQuery().getQuery();
+         if (!(sqlSelectQuery instanceof MySqlSelectQueryBlock)) {
+             // it could be SQLUnionQuery
+             return false;
+         }
+
+        MySqlSelectQueryBlock query = (MySqlSelectQueryBlock) sqlSelectQuery;
+
+        if (!hasGroupByWithOrdinals(query) && !hasOrderByWithOrdinals(query)) {
+            return false;
+        }
+
+        // we cannot clone SQLSelectItem, so we need similar objects to assign to GroupBy and OrderBy items
+        sqlExprGroupCopy = generateAST();
+        sqlExprOrderCopy = generateAST();
+
+        collectOrdinalAliases(sqlExprGroupCopy, groupOrdinalToExpr);
+        collectOrdinalAliases(sqlExprOrderCopy, orderOrdinalToExpr);
+
+        if (groupOrdinalToExpr.isEmpty()) {
+            // orderOrdinalToExpr and groupOrdinalToExpr should be identical
+            // because they are generated from similar syntax trees
+            // checking for groupOrdinalToExpr or orderOrdinalToExpr is enough
+            return false;
+        }
+
+        // we need to rewrite the generated syntax tree to match the original syntax tree Select clause
+        ruleExecutor = RewriteRuleExecutor.builder()
+            .withRule(new UnquoteIdentifierRule())
+            .withRule(new TableAliasPrefixRemoveRule())
+            .build();
+
+        ruleExecutor.executeOn(sqlExprGroupCopy);
+        ruleExecutor.executeOn(sqlExprGroupCopy);
+        return true;
+    }
+
+    @Override
+    public void rewrite(SQLQueryExpr root) {
+        changeOrdinalAliasInGroupAndOrderBy(root);
+    }
+
+    private boolean hasGroupByWithOrdinals(MySqlSelectQueryBlock query) {
+        if (query.getGroupBy() == null) {
+            return false;
+        } else if (query.getGroupBy().getItems().isEmpty()){
+            return false;
+        }
+
+        return query.getGroupBy().getItems().stream().anyMatch(x ->
+            x instanceof MySqlSelectGroupByExpr && ((MySqlSelectGroupByExpr) x).getExpr() instanceof SQLIntegerExpr
+        );
+    }
+
+    private boolean hasOrderByWithOrdinals(MySqlSelectQueryBlock query) {
+        if (query.getOrderBy() == null) {
+            return false;
+        } else if (query.getOrderBy().getItems().isEmpty()){
+            return false;
+        }
+
+        return query.getOrderBy().getItems().stream().anyMatch(x ->
+            x.getExpr() instanceof SQLIntegerExpr
+            || (
+                x.getExpr() instanceof SQLBinaryOpExpr
+                && ((SQLBinaryOpExpr) x.getExpr()).getLeft() instanceof SQLIntegerExpr
+            )
+        );
+    }
+
+    private void collectOrdinalAliases(SQLQueryExpr root, Map<Integer, SQLExpr> map) {
```

Comment:
I checked that `selectList` in Druid (`MySqlSelectQueryBlock`) is an `ArrayList` with getter. So I think this visitor along with a new `Map` is not necessary.